code/espurna/ota_httpupdate.ino

```diff
@@ -0,0 +1,201 @@
+        }
+
+        uint8_t fp_bytes[20] = {0};
+        sslFingerPrintArray(fp_string.c_str(), fp_bytes);
+
+        client->setFingerprint(fp_bytes);
+    }
+
+    // XXX: only for travis. remove when testing against >=2.5.0
+    #if defined(ARDUINO_ESP8266_RELEASE_2_4_2)
+        #define BEARSSL_X509LIST BearSSLX509List
+    #else
+        #define BEARSSL_X509LIST BearSSL::X509List
+    #endif
+
+    BEARSSL_X509LIST *ca = nullptr;
+    if (check == SSL_CLIENT_CHECK_CA) {
+        ca = new BEARSSL_X509LIST(_ota_client_http_update_ca);
+        // because we do not support libc methods of getting time, force client to use ntpclientlib's current time
+        // XXX: local2utc method use is detrimental when DST happening. now() should be utc
+        client->setX509Time(ntpLocal2UTC(now()));
+        client->setTrustAnchors(ca);
+    }
+
+    // TODO: provide terminal command for probing?
+    // TODO: RX and TX buffer sizes must be equal?
+    uint16_t requested_mfln = getSetting("sslMFLN", SSL_CLIENT_MFLN).toInt();
+    if (requested_mfln) {
+        URL _url(url);
+        bool supported = client->probeMaxFragmentLength(_url.host.c_str(), _url.port, requested_mfln);
+        DEBUG_MSG_P(PSTR("[OTA] MFLN buffer size %u supported: %s\n"), requested_mfln, supported ? "yes" : "no");
+        if (supported) {
+            client->setBufferSizes(requested_mfln, requested_mfln);
+        }
+    }
+
+    _otaClientRunUpdater(client.get(), url.c_str());
+
+}
+
+#endif // SSL_CLIENT_BEARSSL
+
+
+#if SSL_CLIENT == SSL_CLIENT_AXTLS
+
+void _otaClientFromHttps(const String& url) {
+
+    int check = getSetting("sslClientCheck", SSL_CLIENT_CHECK_FINGERPRINT).toInt();
+
+    String fp_string = "";
+    if (check == SSL_CLIENT_CHECK_FINGERPRINT) {
+        fp_string = getSetting("otafp", OTA_GITHUB_FP);
+        if (!fp_string.length() || !sslCheckFingerPrint(fp_string.c_str())) {
+            DEBUG_MSG_P(PSTR("[OTA] Wrong fingerprint\n"));
+            return;
+        }
+    }
+
+    _otaClientRunUpdater(nullptr, url, fp_string);
+
+}
+
+#endif // SSL_CLIENT_AXTLS
+
+void _otaClientFrom(const String& url) {
+
+    if (url.startsWith("http://")) {
+        _otaClientFromHttp(url);
+        return;
+    }
+
+    #if SSL_CLIENT_SUPPORT
+        if (url.startsWith("https://")) {
+            _otaClientFromHttps(url);
+            return;
+        }
+    #endif
+
+    DEBUG_MSG_P(PSTR("[OTA] Incorrect URL specified\n"));
+
+}
+
+#if TERMINAL_SUPPORT
+
+void _otaClientInitCommands() {
+
+    terminalRegisterCommand(F("OTA"), [](Embedis* e) {
+        if (e->argc < 2) {
+            terminalError(F("OTA <url>"));
+        } else {
+            _otaClientFrom(String(e->argv[1]));
+            terminalOK();
+        }
+    });
+
+}
+
+#endif // TERMINAL_SUPPORT
+
+// -----------------------------------------------------------------------------
+
+void otaClientSetup() {
+
+    #if TERMINAL_SUPPORT
+        _otaClientInitCommands();
+    #endif
+
```

Comment:
I'm missing the MQTT OTA support here, is that on purpose? Maybe it's better to make the terminal+MQTT code separate of the used OTA module?

Reply:
I did forget that.
Code can be common, but it would need to be a loop() callback so that wificlient works.